sops.go

```diff
@@ -318,142 +320,128 @@ type Store interface {
+	} else {
+		var err error
+		if m.ShamirQuorum == 0 {
+			m.ShamirQuorum = len(m.KeyGroups)
+		}
+		log.Printf("Multiple KeyGroups found, proceeding with Shamir with quorum %d", m.ShamirQuorum)
+		parts, err = shamir.Split(dataKey, len(m.KeyGroups), m.ShamirQuorum)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("Could not split data key into parts for Shamir: %s", err))
+			return
+		}
+		if len(parts) != len(m.KeyGroups) {
+			errs = append(errs, fmt.Errorf("Not enough parts obtained from Shamir. Need %d, got %d", len(m.KeyGroups), len(parts)))
+			return
+		}
+	}
+	for i, group := range m.KeyGroups {
+		part := parts[i]
+		for _, key := range group {
+			svcKey := keyservice.KeyFromMasterKey(key)
+			var keyErrs []error
+			encrypted := false
+			for _, svc := range svcs {
+				rsp, err := svc.Encrypt(context.Background(), &keyservice.EncryptRequest{
+					Key:       &svcKey,
+					Plaintext: part,
+				})
+				if err != nil {
+					keyErrs = append(keyErrs, fmt.Errorf("Failed to encrypt new data key with master key %q: %v\n", key.ToString(), err))
+					continue
+				}
+				key.SetEncryptedDataKey(rsp.Ciphertext)
+				encrypted = true
+				// Only need to encrypt the key successfully with one service
+				break
+			}
+			if !encrypted {
+				errs = append(errs, keyErrs...)
 			}
 		}
 	}
+	m.DataKey = dataKey
 	return
 }
 
 // UpdateMasterKeys encrypts the data key with all master keys
 func (m *Metadata) UpdateMasterKeys(dataKey []byte) (errs []error) {
-	for _, ks := range m.KeySources {
-		for _, k := range ks.Keys {
-			err := k.Encrypt(dataKey)
-			if err != nil {
-				errs = append(errs, fmt.Errorf("Failed to encrypt new data key with master key %q: %v\n", k.ToString(), err))
-			}
-		}
-	}
-	return
+	return m.UpdateMasterKeysWithKeyServices(dataKey, []keyservice.KeyServiceClient{
+		keyservice.NewLocalClient(),
+	})
 }
 
-// AddPGPMasterKeys parses the input comma separated string of GPG fingerprints, generates a PGP MasterKey for each fingerprint, and adds the keys to the PGP KeySource
-func (m *Metadata) AddPGPMasterKeys(pgpFps string) {
-	for i, ks := range m.KeySources {
-		if ks.Name == "pgp" {
-			var keys []MasterKey
-			for _, k := range pgp.MasterKeysFromFingerprintString(pgpFps) {
-				keys = append(keys, k)
-				fmt.Printf("Adding new PGP master key: %X\n", k.Fingerprint)
-			}
-			ks.Keys = append(ks.Keys, keys...)
-			m.KeySources[i] = ks
-		}
+// GetDataKeyWithKeyServices retrieves the data key, asking KeyServices to decrypt it with each
+// MasterKey in the Metadata's KeySources until one of them succeeds.
+func (m Metadata) GetDataKeyWithKeyServices(svcs []keyservice.KeyServiceClient) ([]byte, error) {
+	if m.DataKey != nil {
+		return m.DataKey, nil
 	}
-}
-
-// AddKMSMasterKeys parses the input comma separated string of AWS KMS ARNs, generates a KMS MasterKey for each ARN, and then adds the keys to the KMS KeySource
-func (m *Metadata) AddKMSMasterKeys(kmsArns string, context map[string]*string) {
```

Comment:
This is a niche operation and I think it belongs in the `cmd` package, not in `sops`.